src/crypto/mod.rs

```diff
@@ -52,18 +55,42 @@ impl<E: Engine> PublicKey<E> {
         let w = hash_g1_g2::<E, _>(u, &v).into_affine().mul(r);
         Ciphertext(u, v, w)
     }
+
+    /// Returns a byte string representation of the public key.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.0.into_affine().into_compressed().as_ref().to_vec()
+    }
 }
 
 /// A signature, or a signature share.
-#[derive(Debug)]
+#[derive(Clone, PartialOrd)]
 pub struct Signature<E: Engine>(E::G2);
 
+impl<E: Engine> fmt::Debug for Signature<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let uncomp = self.0.into_affine().into_uncompressed();
+        let bytes = uncomp.as_ref();
+        write!(f, "{:?}", HexBytes(bytes))
+    }
+}
+
 impl<E: Engine> PartialEq for Signature<E> {
     fn eq(&self, other: &Signature<E>) -> bool {
         self.0 == other.0
     }
 }
 
+impl<E: Engine> Signature<E> {
+    pub fn parity(&self) -> bool {
+        let uncomp = self.0.into_affine().into_uncompressed();
```

Comment:
I wonder whether that's unique, i.e. does `a == b` imply `a.into_affine().into_uncompressed() == b.into_affine().into_uncompressed()`? (I assume it does, but I'm not sure.) Even if that holds true for `Bls12` but is not a guarantee of `Engine` in general, I'd maybe make this method `Bls12`-only.

Reply:
No, I'm worried about [`E::G2`'s `PartialEq` implementation](https://github.com/ebfull/pairing/blob/master/src/bls12_381/ec.rs#L45) itself.
I'm wondering whether each node will compute the same value of `uncomp` here: If there are different valid in-memory representations for the same point on the affine curve, it could happen that `a == b`, even though they are not bitwise equal. ([Projective coordinates](https://en.wikipedia.org/wiki/Elliptic-curve_cryptography#Projective_coordinates) are not unique.) The question is whether the same point can even have two different valid compressed representations.
But I think it can't, so we should be fine: The `pairing` documentation doesn't explain this but it looks like it is just [the x coordinate plus a few flags](https://github.com/ebfull/pairing/blob/master/src/bls12_381/ec.rs#L1510), which should all be unique.

Reply:
The wikipedia article call this system Jacobian as opposed to projective.

Reply:
This `PartialEq` instance is only required for a few assertion tests at the moment where coincidence of two points via their projective coordinates would be highly unlikely. I tried to introduce conditional compilation with `#[cfg(debug_assertions)] but that made the instance not compile in test `--release`  configuration where the debug assertions are used.
An alternative could be making the private field of `Signature` accessible from outside the module by a method or just by making the field public, and then defining the equality relation in the tests, where it does not interfere with the library.

Reply:
My concern was not the `PartialEq`, but whether the coin is actually common: If the uncompressed representations that two nodes compute (even though they refer to the same point of the curve) have different bytes, then they might get different coin values.
I'm pretty sure that can't happen, and they _are_ guaranteed to be equal, but the `pairing` docs don't say that explicitly.

Reply:
Actually, it looks like for `G2Affine` (and `Fq2`, `Fq` and `FqRepr`), `PartialEq` is derived, so they do seem to have unique representations. This should be fine then. :+1: 
(I think `x.c0 + x.c1 * u` _is_ unique, because `u` is not in the base field `Fq`.)

Reply:
Glad we looked into it! If so, we are done with this PR?